src/actions.rs

```diff
@@ -77,6 +77,64 @@ impl ActionHandler {
             pub spans: Vec<span::compiler::DiagnosticSpan>,
         }
 
+        fn parse_compiler_messages(messages: &Vec<String>,
+                                   results: &mut HashMap<PathBuf, Vec<Diagnostic>>) {
+            for msg in messages {
+                let message = match serde_json::from_str::<CompilerMessage>(&msg) {
+                    Ok(message) => message,
+                    Err(e) => {
+                        debug!("build error {:?}", e);
+                        debug!("from {}", msg);
+                        continue;
+                    }
+                };
+
+                if message.spans.is_empty() {
+                    continue;
+                }
+
+                let span = message.spans[0].rls_span().zero_indexed();
+
+                let diag = Diagnostic {
+                    range: ls_util::rls_to_range(span.range),
+                    severity: Some(if message.level == "error" {
+                        DiagnosticSeverity::Error
+                    } else {
+                        DiagnosticSeverity::Warning
+                    }),
+                    code: Some(NumberOrString::String(match message.code {
+                        Some(c) => c.code.clone(),
+                        None => String::new(),
+                    })),
+                    source: Some("rustc".into()),
+                    message: message.message.clone(),
+                };
+
+                let mut results = results.entry(span.file.clone()).or_insert(vec![]);
+
+                results.push(diag);
+            }
+        }
+
+        fn convert_build_results_to_notifications(
+            build_results: &HashMap<PathBuf, Vec<Diagnostic>>,
+            project_path: &Path
+        ) -> Vec<NotificationMessage<PublishDiagnosticsParams>> {
```

Comment:
I think it fits into exactly 100 chars, so it is fine (tbh, we exceed the 100 char limit in a few places already, so I don't mind going over a bit in any case).